File: zroky-sdk/zroky/_verify.py
# SPDX-License-Identifier: FSL-1.1-MIT
# Copyright 2026 Zroky AI

"""Saved connector outcome verification helpers."""
from __future__ import annotations

from typing import Any, Literal

import httpx

from zroky._errors import ZrokyOutcomeVerificationError

_REQUEST_TIMEOUT_S = 15.0

SavedVerificationConnector = Literal["generic_rest", "ledger_refund", "crm_record"]


def _outcome_reconciliation_url(ingest_url: str, connector: SavedVerificationConnector) -> str:
    paths = {
        "generic_rest": "/v1/outcomes/reconciliation/generic-rest/saved",
        "ledger_refund": "/v1/outcomes/reconciliation/ledger-refund/saved",
        "crm_record": "/v1/outcomes/reconciliation/customer-record/saved",
    }
    return ingest_url.rstrip("/") + paths[connector]


def _connector_payload(
    *,
    connector: SavedVerificationConnector,
    record_ref: str | None,
    refund_id: str | None,
    customer_id: str | None,
) -> dict[str, Any]:
    if connector == "generic_rest":
        cleaned_record_ref = (record_ref or "").strip()
        if not cleaned_record_ref:
            raise ZrokyOutcomeVerificationError(
                "[ZROKY] Generic REST outcome verification requires record_ref."
            )
        return {"record_ref": cleaned_record_ref}
    if connector == "ledger_refund":
        return {"refund_id": refund_id} if refund_id else {}
    if connector == "crm_record":
        return {"customer_id": customer_id} if customer_id else {}
    raise ZrokyOutcomeVerificationError(
        f"[ZROKY] Unsupported verification connector: {connector!r}."
    )


def verify_outcome(
    *,
    connector: SavedVerificationConnector,
    claimed: dict[str, Any],
    record_ref: str | None = None,
    refund_id: str | None = None,
    customer_id: str | None = None,
    call_id: str | None = None,
    trace_id: str | None = None,
    runtime_policy_decision_id: str | None = None,
    action_type: str | None = None,
    system_ref: str | None = None,
    match_fields: list[str] | tuple[str, ...] | None = None,
    amount_usd: float | None = None,
    currency: str | None = None,
    idempotency_key: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """Verify an agent's claimed post-action outcome through a saved connector.

    The SDK sends only the record reference and claimed result. Customer-owned
    connector credentials remain stored server-side in Zroky.
    """
    import zroky as _z

    if _z._config is None:
        _z.init()
    cfg = _z._config
    if cfg is None:
        raise ZrokyOutcomeVerificationError("[ZROKY] Outcome verification is not initialized.")
    if not cfg.api_key or not cfg.project:
        raise ZrokyOutcomeVerificationError(
            "[ZROKY] Outcome verification requires api_key and project."
        )

    payload: dict[str, Any] = {
        **_connector_payload(
            connector=connector,
            record_ref=record_ref,
            refund_id=refund_id,
            customer_id=customer_id,
        ),
        "call_id": call_id,
        "trace_id": trace_id,
        "runtime_policy_decision_id": runtime_policy_decision_id,
        "action_type": action_type,
        "system_ref": system_ref,
        "claimed": claimed,
        "match_fields": list(match_fields) if match_fields else None,
        "amount_usd": amount_usd,
        "currency": currency,
        "idempotency_key": idempotency_key,
        "metadata": metadata,
    }
    payload = {key: value for key, value in payload.items() if value is not None}

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {cfg.api_key}",
        "x-api-key": cfg.api_key,
        "x-project-id": cfg.project,
    }

    try:
        response = httpx.post(
            _outcome_reconciliation_url(cfg.ingest_url, connector),
            headers=headers,
            json=payload,
            timeout=_REQUEST_TIMEOUT_S,
        )
    except httpx.HTTPError as exc:
        raise ZrokyOutcomeVerificationError(
            f"[ZROKY] Outcome verification unavailable: {exc}",
        ) from exc

    if response.status_code >= 300:
        raise ZrokyOutcomeVerificationError(
            f"[ZROKY] Outcome verification failed with HTTP {response.status_code}."
        )

    try:
        result = response.json()
    except ValueError as exc:
        raise ZrokyOutcomeVerificationError(
            f"[ZROKY] Outcome verification returned invalid JSON: {exc}",
        ) from exc
    if not isinstance(result, dict):
        raise ZrokyOutcomeVerificationError(
            "[ZROKY] Outcome verification returned an invalid response shape."
        )
    return result

File: zroky-sdk-js/src/verify.ts
// SPDX-License-Identifier: FSL-1.1-MIT
// Copyright 2026 Zroky AI

import { nodeEnv, resolveApiBase } from "./api";
import { resolveConfig } from "./config";
import type { ZrokyJsonValue, ZrokyRiskActionType } from "./contracts";
import type { ZrokyConfig } from "./types";

export type SavedVerificationConnector = "generic_rest" | "ledger_refund" | "crm_record";
export type OutcomeVerificationVerdict = "matched" | "mismatched" | "not_verified";

export interface OutcomeReconciliationView {
  id: string;
  project_id: string;
  call_id: string | null;
  trace_id: string | null;
  runtime_policy_decision_id: string | null;
  action_type: string | null;
  connector_type: string;
  system_ref: string | null;
  verdict: OutcomeVerificationVerdict;
  reason: string | null;
  amount_usd: number | null;
  currency: string | null;
  claimed: Record<string, unknown>;
  actual: Record<string, unknown> | null;
  comparison: Record<string, unknown>;
  idempotency_key: string | null;
  metadata: Record<string, unknown> | null;
  checked_at: string;
  created_at: string;
}

interface VerifyOutcomeBaseOptions {
  connector: SavedVerificationConnector;
  claimed: Record<string, unknown>;
  callId?: string;
  traceId?: string;
  runtimePolicyDecisionId?: string;
  actionType?: ZrokyRiskActionType | (string & {});
  systemRef?: string;
  matchFields?: string[];
  amountUsd?: number;
  currency?: string;
  idempotencyKey?: string;
  metadata?: Record<string, ZrokyJsonValue | unknown>;
}

export interface VerifyGenericRestOutcomeOptions extends VerifyOutcomeBaseOptions {
  connector: "generic_rest";
  recordRef: string;
}

export interface VerifyLedgerRefundOutcomeOptions extends VerifyOutcomeBaseOptions {
  connector: "ledger_refund";
  refundId?: string;
}

export interface VerifyCrmRecordOutcomeOptions extends VerifyOutcomeBaseOptions {
  connector: "crm_record";
  customerId?: string;
}

export type VerifyOutcomeOptions =
  | VerifyGenericRestOutcomeOptions
  | VerifyLedgerRefundOutcomeOptions
  | VerifyCrmRecordOutcomeOptions;

export class ZrokyOutcomeVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ZrokyOutcomeVerificationError";
  }
}

function withoutUndefined(value: Record<string, unknown>): Record<string, unknown> {
  return Object.fromEntries(Object.entries(value).filter(([, item]) => item !== undefined));
}

function endpointPath(connector: SavedVerificationConnector): string {
  switch (connector) {
    case "generic_rest":
      return "/v1/outcomes/reconciliation/generic-rest/saved";
    case "ledger_refund":
      return "/v1/outcomes/reconciliation/ledger-refund/saved";
    case "crm_record":
      return "/v1/outcomes/reconciliation/customer-record/saved";
  }
}

function connectorSpecificPayload(options: VerifyOutcomeOptions): Record<string, unknown> {
  switch (options.connector) {
    case "generic_rest":
      return { record_ref: options.recordRef };
    case "ledger_refund":
      return { refund_id: options.refundId };
    case "crm_record":
      return { customer_id: options.customerId };
  }
}

function verificationPayload(options: VerifyOutcomeOptions): Record<string, unknown> {
  return withoutUndefined({
    ...connectorSpecificPayload(options),
    call_id: options.callId,
    trace_id: options.traceId,
    runtime_policy_decision_id: options.runtimePolicyDecisionId,
    action_type: options.actionType,
    system_ref: options.systemRef,
    claimed: options.claimed,
    match_fields: options.matchFields,
    amount_usd: options.amountUsd,
    currency: options.currency,
    idempotency_key: options.idempotencyKey,
    metadata: options.metadata,
  });
}

/**
 * Verify an agent's claimed post-action outcome against a saved Zroky connector.
 *
 * This calls Zroky's saved connector runtime. Customer connector credentials stay
 * server-side; the SDK only sends the record reference and claimed outcome.
 */
export async function verifyOutcome(
  options: VerifyOutcomeOptions,
  config: ZrokyConfig = {},
): Promise<OutcomeReconciliationView> {
  const resolved = resolveConfig(config);
  if (resolved.disabled) {
    throw new ZrokyOutcomeVerificationError("[ZROKY] Outcome verification is disabled.");
  }

  const env = nodeEnv();
  const apiKey = resolved.apiKey ?? env?.["ZROKY_API_KEY"];
  const projectId = resolved.projectId ?? env?.["ZROKY_PROJECT_ID"] ?? env?.["ZROKY_PROJECT"];
  if (!apiKey || !projectId) {
    throw new ZrokyOutcomeVerificationError(
      "[ZROKY] Outcome verification requires apiKey and projectId.",
    );
  }

  const endpoint = resolveApiBase(resolved.endpoint ?? env?.["ZROKY_API_URL"] ?? env?.["ZROKY_ENDPOINT"]);
  const url = `${endpoint}${endpointPath(options.connector)}`;

  let response: Response;
  try {
    response = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": apiKey,
        "x-project-id": projectId,
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(verificationPayload(options)),
    });
  } catch (error) {
    throw new ZrokyOutcomeVerificationError(
      `[ZROKY] Outcome verification unavailable: ${String(
        (error as { message?: unknown })?.message ?? error,
      )}`,
    );
  }

  if (!response.ok) {
    throw new ZrokyOutcomeVerificationError(
      `[ZROKY] Outcome verification failed with HTTP ${response.status}.`,
    );
  }

  try {
    return (await response.json()) as OutcomeReconciliationView;
  } catch (error) {
    throw new ZrokyOutcomeVerificationError(
      `[ZROKY] Outcome verification returned invalid JSON: ${String(
        (error as { message?: unknown })?.message ?? error,
      )}`,
    );
  }
}
